fdbclient/VersionVector.h

```diff
@@ -35,24 +35,39 @@ struct VersionVector {
 	                    // there may or may not be a corresponding entry for this
 	                    // version in the "versions" map.)
 
-	VersionVector() : maxVersion(invalidVersion) {}
-	VersionVector(Version version) : maxVersion(version) {}
+	VersionVector() : maxVersion(invalidVersion), cachedEncodedSize(0) {}
+	VersionVector(Version version) : maxVersion(version), cachedEncodedSize(0) {}
 
 private:
 	// Only invoked by getDelta() and applyDelta(), where tag has been validated
 	// and version is guaranteed to be larger than the existing value.
-	inline void setVersionNoCheck(const Tag& tag, Version version) { versions[tag] = version; }
+	inline void setVersionNoCheck(const Tag& tag, Version version) {
+		versions[tag] = version;
+		invalidateCachedEncodedSize();
+	}
+
+	// Encoded version vector size. Introduced to help speed up serialization.
+	// @note This encoded size is not meant to be kept in sync with the updates
+	// that happen to the version vector.
+	// @note A value of 0 indicates that the encoded version vector size is not
+	// cached.
+	size_t cachedEncodedSize;
 
 public:
 	Version getMaxVersion() const { return maxVersion; }
 
+	void setMaxVersion(Version version) { maxVersion = version; }
```

Comment:
Need `invalidateCachedEncodedSize()` here.

Reply:
It's good you noticed it. I didn't invalidate here because setting/updating "maxVersion" doesn't affect the number of bytes that are needed to capture the encoded version vector (going by the calculations in "getEncodedSize()").

Reply:
Well, we can invalidate the cache here too, if you think that would be clearer.

Reply:
Isn't `maxVersion` going to change the encoded size? Because the version may need to be changed from e.g., 16 bits to 32 bits?

Reply:
No, we are always reserving "sizeof(Version)" to serialize "VersionVector::maxSize":

> sizeof(Version); /* VersionVector::maxVersion */

Reply:
I think you are right, here we don't need `invalidateCachedEncodedSize()`.